modules/cas_cache/layer_cache_management.c

```diff
@@ -15,141 +15,367 @@ extern u32 seq_cut_off_mb;
+	result = wait_for_completion_interruptible(&context->compl);
+
+	if (result < 0)
+		result = -KCAS_ERR_WAITING_INTERRUPTED;
+	else
+		result = context->result;
+
+	if (atomic_dec_and_test(&context->ref_count))
+		kfree(context);
 
-	atomic_set(&cache_priv->flush_interrupt_enabled, 0);
-	ocf_mngt_core_flush(core, _cache_mngt_core_flush_complete, &context);
-	wait_for_completion_interruptible(&context.compl);
 	atomic_set(&cache_priv->flush_interrupt_enabled, 1);
 
 	return result;
 }
 
+static void _cache_mngt_cache_priv_deinit(ocf_cache_t cache)
+{
+	struct cache_priv *cache_priv = ocf_cache_get_priv(cache);
+
+	vfree(cache_priv);
+}
+
+static int exit_instance_finish(ocf_cache_t cache, int error)
+{
+	struct cache_priv *cache_priv;
+	ocf_queue_t mngt_queue;
+	bool flush_status;
+
+	flush_status = ocf_mngt_cache_is_dirty(cache);
+	cache_priv = ocf_cache_get_priv(cache);
+	mngt_queue = cache_priv->mngt_queue;
+
+	if (error && error != -OCF_ERR_WRITE_CACHE)
+		goto restore_exp_obj;
+
+	if (!error && flush_status)
+		error = -KCAS_ERR_STOPPED_DIRTY;
+
+	module_put(THIS_MODULE);
+
+	cas_cls_deinit(cache);
+
+	vfree(cache_priv);
+
+	ocf_mngt_cache_unlock(cache);
+	ocf_mngt_cache_put(cache);
+	ocf_queue_put(mngt_queue);
+
+	return error;
+
+restore_exp_obj:
+	if (block_dev_create_all_exported_objects(cache)) {
+		/* Print error msg but do not change return err code to inform user why
+		* stop failed originally. */
+		printk(KERN_WARNING
+			"Failed to restore (create) all exported objects!\n");
+		goto unlock;
+	}
+	if (block_dev_activate_all_exported_objects(cache)) {
+		block_dev_destroy_all_exported_objects(cache);
+		printk(KERN_WARNING
+			"Failed to restore (activate) all exported objects!\n");
+	}
+unlock:
+	ocf_mngt_cache_unlock(cache);
+	ocf_mngt_cache_put(cache);
+	return error;
+}
+
+struct _cache_mngt_attach_context {
+	struct completion compl;
+	struct kcas_start_cache *cmd;
+	struct ocf_mngt_cache_device_config *device_cfg;
+	int result;
+	spinlock_t start_sync;
+
+	struct {
+		bool priv_inited:1;
+		bool cls_inited:1;
+	};
+};
+
+static void cache_start_rollback(struct _cache_mngt_attach_context *ctx,
+		ocf_cache_t cache)
+{
+	struct cache_priv *cache_priv;
+	ocf_queue_t mngt_queue;
+
+	cache_priv = ocf_cache_get_priv(cache);
+	mngt_queue = cache_priv->mngt_queue;
+
+	if (ctx->cls_inited)
+		cas_cls_deinit(cache);
+
+	if (ctx->priv_inited)
+		_cache_mngt_cache_priv_deinit(cache);
+
+	if (mngt_queue)
+		ocf_queue_put(mngt_queue);
+
+	ocf_mngt_cache_unlock(cache);
+
+	module_put(THIS_MODULE);
+}
+
+static void _cache_mngt_cache_stop_rollback_complete(ocf_cache_t cache,
+		void *priv, int error)
+{
+	struct _cache_mngt_attach_context *ctx = priv;
+
+	cache_start_rollback(ctx, cache);
+
+	complete(&ctx->compl);
+}
+
 static void _cache_mngt_cache_stop_complete(ocf_cache_t cache, void *priv,
 		int error)
 {
-	struct _cache_mngt_sync_context *context = priv;
+	struct _cache_mngt_async_lock_context *context = priv;
+	bool interrupt;
```

Comment:
I suggest renaming interrupt->interrupted. "interrupt" is unfortunate, as completion is often called in interrupt context.